File: app/src/main/java/com/club/medlems/data/entity/Entities.kt
package com.club.medlems.data.entity

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate

/** Member lifecycle stage */
enum class MemberType {
    /** Registered on tablet, no membershipId assigned yet */
    TRIAL,
    /** Has official membershipId assigned */
    FULL
}

/** Member operational status */
enum class MemberStatus { ACTIVE, INACTIVE }
enum class PracticeType { Riffel, Pistol, LuftRiffel, LuftPistol, Andet }
enum class TrainerLevel { FULL, ASSISTANT }
enum class ScanEventType { FIRST_SCAN, REPEAT_SCAN }
enum class SessionSource { kiosk, attendant }

/** Registration approval status for new member registrations - DEPRECATED: Use MemberType instead */
@Deprecated("Use MemberType instead. Will be removed after migration.")
enum class ApprovalStatus { PENDING, APPROVED, REJECTED }

@Entity(
    indices = [
        Index(value = ["membershipId"], unique = true),
        Index(value = ["memberType"]),
        Index(value = ["status"]),
        Index(value = ["lastName", "firstName"])
    ]
)
data class Member(
    /** Immutable UUID, primary key across all devices */
    @PrimaryKey 
    val internalId: String,
    
    /** Club-assigned ID, null for trial members */
    val membershipId: String? = null,
    
    /** Lifecycle stage: TRIAL or FULL */
    val memberType: MemberType = MemberType.TRIAL,
    
    /** Operational status: ACTIVE or INACTIVE */
    val status: MemberStatus = MemberStatus.ACTIVE,
    
    // === Personal Information ===
    val firstName: String,
    val lastName: String,
    val birthDate: LocalDate? = null,
    val gender: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val address: String? = null,
    val zipCode: String? = null,
    val city: String? = null,
    
    // === Guardian Information (for minors) ===
    val guardianName: String? = null,
    val guardianPhone: String? = null,
    val guardianEmail: String? = null,
    
    // === Membership Details ===
    val expiresOn: String? = null, // ISO local date string for simplicity
    val registrationPhotoPath: String? = null,

    // === ID Photo for Adult Verification (Enhanced Trial Registration) ===
    /** Path to ID photo file on disk (adults only, for verification) */
    val idPhotoPath: String? = null,

    // === Merge Tracking (per DD-10) ===
    /** If merged into another member, points to surviving member's internalId */
    val mergedIntoId: String? = null,
    
    // === Timestamps ===
    val createdAtUtc: Instant,
    val updatedAtUtc: Instant = Instant.DISTANT_PAST,
    
    // === Sync Metadata ===
    /** Device that created/last modified this record */
    val deviceId: String? = null,
    /** Monotonically increasing version for conflict detection */
    val syncVersion: Long = 0,
    /** Last successful sync timestamp */
    val syncedAtUtc: Instant? = null
)

@Entity(indices = [
    androidx.room.Index(value = ["internalMemberId", "localDate"]),
    androidx.room.Index(value = ["membershipId", "localDate"])
])
data class CheckIn(
    @PrimaryKey val id: String,
    /** FK to Member.internalId - the primary member reference */
    val internalMemberId: String,
    /** @deprecated Use internalMemberId. Kept for backward compatibility with older sync. */
    @Deprecated("Use internalMemberId instead")
    val membershipId: String? = null,
    val createdAtUtc: Instant,
    val localDate: LocalDate,
    val firstOfDayFlag: Boolean = true,
    
    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

@Entity(indices = [
    androidx.room.Index(value = ["internalMemberId"]),
    androidx.room.Index(value = ["localDate"]),
    androidx.room.Index(value = ["practiceType", "localDate"]),
    androidx.room.Index(value = ["internalMemberId", "practiceType", "classification"])
])
data class PracticeSession(
    @PrimaryKey val id: String,
    /** FK to Member.internalId - the primary member reference */
    val internalMemberId: String,
    /** @deprecated Use internalMemberId. Kept for backward compatibility. */
    @Deprecated("Use internalMemberId instead")
    val membershipId: String? = null,
    val createdAtUtc: Instant,
    val localDate: LocalDate,
    val practiceType: PracticeType,
    val points: Int,
    val krydser: Int?,
    val classification: String? = null,
    val source: SessionSource,
    
    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

@Entity(indices = [
    androidx.room.Index(value = ["internalMemberId", "createdAtUtc"])
])
data class ScanEvent(
    @PrimaryKey val id: String,
    /** FK to Member.internalId - the primary member reference */
    val internalMemberId: String,
    /** @deprecated Use internalMemberId. Kept for backward compatibility. */
    @Deprecated("Use internalMemberId instead")
    val membershipId: String? = null,
    val createdAtUtc: Instant,
    val type: ScanEventType,
    val linkedCheckInId: String? = null,
    val linkedSessionId: String? = null,
    val canceledFlag: Boolean = false,
    
    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

@Entity
data class NewMemberRegistration(
    @PrimaryKey val id: String,
    val temporaryId: String,
    val createdAtUtc: Instant,
    val photoPath: String,
    val firstName: String,
    val lastName: String,
    val email: String? = null,
    val phone: String? = null,
    val birthDate: String? = null,
    val gender: String? = null,
    val address: String? = null,
    val zipCode: String? = null,
    val city: String? = null,
    val guardianName: String? = null,
    val guardianPhone: String? = null,
    val guardianEmail: String? = null,
    
    // Approval workflow fields
    val approvalStatus: ApprovalStatus = ApprovalStatus.PENDING,
    val approvedAtUtc: Instant? = null,
    val rejectedAtUtc: Instant? = null,
    val rejectionReason: String? = null,
    val createdMemberId: String? = null,
    
    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

// ===== Equipment Management Entities (Phase 3 - Trainer Tablet) =====

/** Equipment item status */
enum class EquipmentStatus { 
    /** Available for checkout */
    Available, 
    /** Currently checked out to a member */
    CheckedOut, 
    /** Under maintenance, not available */
    Maintenance, 
    /** No longer in use */
    Retired 
}

/** Equipment type category */
enum class EquipmentType { 
    /** Training materials (targets, stands, etc.) */
    TrainingMaterial 
}

/** Conflict resolution status for equipment checkouts */
enum class ConflictStatus { 
    /** Conflict detected, awaiting resolution */
    Pending, 
    /** Conflict has been resolved */
    Resolved, 
    /** Conflicting checkout was cancelled */
    Cancelled 
}

/**
 * Equipment item entity for tracking club equipment.
 * 
 * @see [design.md FR-8.2] - EquipmentItem schema
 */
@Entity(indices = [
    androidx.room.Index(value = ["serialNumber"], unique = true),
    androidx.room.Index(value = ["status"])
])
data class EquipmentItem(
    @PrimaryKey val id: String,
    /** Human-readable identifier, manually entered */
    val serialNumber: String,
    /** Equipment category */
    val type: EquipmentType = EquipmentType.TrainingMaterial,
    /** Optional description, max 200 characters */
    val description: String? = null,
    /** Current status */
    val status: EquipmentStatus = EquipmentStatus.Available,
    /** Optional discipline this equipment is associated with */
    val discipline: PracticeType? = null,
    /** Device that created this equipment item */
    val createdByDeviceId: String,
    val createdAtUtc: Instant,
    val modifiedAtUtc: Instant,

    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

/**
 * Equipment checkout record for tracking equipment loans to members.
 *
 * @see [design.md FR-8.3] - EquipmentCheckout schema
 */
@Entity(indices = [
    androidx.room.Index(value = ["equipmentId"]),
    androidx.room.Index(value = ["internalMemberId"]),
    androidx.room.Index(value = ["membershipId"]),
    androidx.room.Index(value = ["checkedInAtUtc"]),
    androidx.room.Index(value = ["conflictStatus"])
])
data class EquipmentCheckout(
    @PrimaryKey val id: String,
    /** Foreign key to EquipmentItem */
    val equipmentId: String,
    /** FK to Member.internalId - the primary member reference */
    val internalMemberId: String,
    /** @deprecated Use internalMemberId. Kept for backward compatibility. */
    @Deprecated("Use internalMemberId instead")
    val membershipId: String? = null,
    /** Checkout timestamp */
    val checkedOutAtUtc: Instant,
    /** Return timestamp (null if still checked out) */
    val checkedInAtUtc: Instant? = null,
    /** Device that performed checkout */
    val checkedOutByDeviceId: String,
    /** Device that performed check-in (null if still out) */
    val checkedInByDeviceId: String? = null,
    /** Optional notes at checkout, max 500 characters */
    val checkoutNotes: String? = null,
    /** Optional notes at return, max 500 characters */
    val checkinNotes: String? = null,
    /** Conflict status for offline checkout conflicts */
    val conflictStatus: ConflictStatus? = null,
    /** Resolution notes if conflict was resolved */
    val conflictResolutionNotes: String? = null,
    val createdAtUtc: Instant,
    val modifiedAtUtc: Instant,

    // Sync metadata fields
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

// ===== Member Preference Entity (for sync of UI preferences) =====

/**
 * Member practice preferences (last selected discipline/classification).
 * Synced between tablets via the laptop to preserve preferences when replacing devices.
 *
 * @see [design.md member-preference-sync] - Member Preference Sync feature
 */
@Entity(tableName = "member_preference")
data class MemberPreference(
    /** FK to Member.internalId */
    @PrimaryKey
    val memberId: String,

    /** Last selected PracticeType enum name (e.g., "Riffel", "Pistol") */
    val lastPracticeType: String? = null,

    /** Last selected classification within the practice type */
    val lastClassification: String? = null,

    /** When this preference was last updated */
    val updatedAtUtc: Instant
)

// ===== Trainer Experience Entities =====

/**
 * Trainer information for a member.
 * Tracks whether a member is a trainer and their certifications.
 *
 * @see [trainer-experience/prd.md] - Trainer Experience Feature
 */
@Entity(
    tableName = "trainer_info",
    foreignKeys = [ForeignKey(
        entity = Member::class,
        parentColumns = ["internalId"],
        childColumns = ["memberId"],
        onDelete = ForeignKey.CASCADE
    )]
)
data class TrainerInfo(
    /** FK to Member.internalId */
    @PrimaryKey
    val memberId: String,

    /** Whether the member is designated as a trainer */
    val isTrainer: Boolean = false,

    /** Whether the member has Skydeleder (Range Officer) certification */
    val hasSkydelederCertificate: Boolean = false,

    /** Date when Skydeleder certificate was obtained */
    val certifiedDate: Instant? = null,

    // Timestamps
    val createdAtUtc: Instant,
    val modifiedAtUtc: Instant,

    // Sync metadata
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

/**
 * Trainer discipline qualification.
 * Tracks which disciplines a trainer is qualified to supervise.
 *
 * @see [trainer-experience/prd.md] - Trainer Experience Feature
 */
@Entity(
    tableName = "trainer_discipline",
    foreignKeys = [ForeignKey(
        entity = Member::class,
        parentColumns = ["internalId"],
        childColumns = ["memberId"],
        onDelete = ForeignKey.CASCADE
    )],
    indices = [Index("memberId")]
)
data class TrainerDiscipline(
    @PrimaryKey
    val id: String,

    /** FK to Member.internalId */
    val memberId: String,

    /** The discipline type (Riffel, Pistol, etc.) */
    val discipline: PracticeType,

    /** Trainer level for this discipline */
    val level: TrainerLevel,

    /** Date when certification for this discipline was obtained */
    val certifiedDate: Instant? = null,

    // Timestamps
    val createdAtUtc: Instant,
    val modifiedAtUtc: Instant,

    // Sync metadata
    val deviceId: String? = null,
    val syncVersion: Long = 0,
    val syncedAtUtc: Instant? = null
)

File: laptop/src/types/entities.ts
/**
 * Type definitions matching Android Room entities.
 * These types mirror the Kotlin data classes in the Android app.
 * 
 * @see /app/src/main/java/com/club/medlems/data/entity/Entities.kt
 */

// ===== Member Types =====

/** Member lifecycle stage: TRIAL (no membershipId) or FULL (has membershipId) */
export type MemberLifecycleStage = 'TRIAL' | 'FULL';

/** Member operational status */
export type MemberStatus = 'ACTIVE' | 'INACTIVE';
export type Gender = 'MALE' | 'FEMALE' | 'OTHER';

/** Fee category for membership dues calculation */
export type FeeCategoryType = 'ADULT' | 'CHILD' | 'CHILD_PLUS' | 'HONORARY';

export interface Member {
  /** Immutable UUID, primary key across all devices */
  internalId: string;

  /** Club-assigned ID, null for trial members */
  membershipId: string | null;

  /** Lifecycle stage: TRIAL or FULL */
  memberLifecycleStage: MemberLifecycleStage;

  /** Operational status: ACTIVE or INACTIVE */
  status: MemberStatus;

  // Personal Information
  firstName: string;
  lastName: string;
  birthDate: string | null; // ISO date string YYYY-MM-DD (DB column name)
  gender: Gender | null;
  email: string | null;
  phone: string | null;
  address: string | null;
  zipCode: string | null;
  city: string | null;

  // Guardian info for members under 18
  guardianName: string | null;
  guardianPhone: string | null;
  guardianEmail: string | null;

  // Member type for fee calculation (ADULT/CHILD/CHILD_PLUS) - DB column: memberType
  memberType: FeeCategoryType;

  // Membership details
  expiresOn: string | null;
  /** @deprecated Use photoPath and photoThumbnail instead */
  registrationPhotoPath: string | null;
  /** Path to full-resolution photo file on disk */
  photoPath: string | null;
  /** Small 150x150 thumbnail as data URL for list views */
  photoThumbnail: string | null;

  // ID photo for adult verification (Enhanced Trial Registration)
  /** Path to ID photo file on disk (adults only) */
  idPhotoPath: string | null;
  /** Small 150x150 ID photo thumbnail as data URL */
  idPhotoThumbnail: string | null;

  // Merge tracking (per DD-10)
  mergedIntoId: string | null;

  // Timestamps
  createdAtUtc: string; // ISO datetime
  updatedAtUtc: string; // ISO datetime

  // Sync metadata
  syncedAtUtc: string | null;
  syncVersion: number;
}

// NOTE: MemberForTabletSync removed per DD-9
// All sync operations now use full Member type directly

/**
 * Lightweight member type for list views.
 * Only includes fields needed for rendering member lists.
 * Uses photoThumbnail instead of full photo for performance.
 */
export interface MemberListItem {
  internalId: string;
  membershipId: string | null;
  memberLifecycleStage: MemberLifecycleStage;
  status: MemberStatus;
  firstName: string;
  lastName: string;
  birthDate: string | null;
  photoThumbnail: string | null;
  /** ID photo thumbnail for showing ID status in lists */
  idPhotoThumbnail: string | null;
  createdAtUtc: string;
}

// ===== Check-in Types =====

export interface CheckIn {
  id: string;
  internalMemberId: string; // FK to Member.internalId
  /** @deprecated Use internalMemberId. Retained for backward compatibility. */
  membershipId: string | null;
  localDate: string; // ISO date YYYY-MM-DD
  createdAtUtc: string;
  syncedAtUtc: string | null;
  syncVersion: number;
}

// ===== Practice Session Types =====

export type PracticeType = 'RIFLE' | 'PISTOL';

export interface PracticeSession {
  id: string;
  internalMemberId: string; // FK to Member.internalId
  /** @deprecated Use internalMemberId. Retained for backward compatibility. */
  membershipId: string | null;
  localDate: string;
  practiceType: PracticeType;
  classification: string;
  points: number;
  krydser: number | null;
  notes: string | null;
  createdAtUtc: string;
  syncedAtUtc: string | null;
  syncVersion: number;
}

// ===== Scan Event Types =====

export type ScanType = 'FIRST_SCAN' | 'REPEAT_SCAN';

export interface ScanEvent {
  id: string;
  internalMemberId: string; // FK to Member.internalId
  /** @deprecated Use internalMemberId. Retained for backward compatibility. */
  membershipId: string | null;
  scanType: ScanType;
  linkedCheckInId: string | null;
  linkedSessionId: string | null;
  canceledFlag: boolean;
  createdAtUtc: string;
  syncedAtUtc: string | null;
  syncVersion: number;
}

// ===== New Member Registration Types =====

export type ApprovalStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

export interface NewMemberRegistration {
  id: string;
  firstName: string;
  lastName: string;
  birthday: string | null;
  gender: Gender | null;
  email: string | null;
  phone: string | null;
  address: string | null;
  zipCode: string | null;
  city: string | null;
  notes: string | null;
  // Photo path - stored locally after sync receives base64 data
  photoPath: string | null;
  // Guardian info for under-18 registrations
  guardianName: string | null;
  guardianPhone: string | null;
  guardianEmail: string | null;
  // Source device info
  sourceDeviceId: string;
  sourceDeviceName: string | null;
  // Approval workflow
  approvalStatus: ApprovalStatus;
  approvedAtUtc: string | null;
  rejectedAtUtc: string | null;
  rejectionReason: string | null;
  createdMemberId: string | null;
  createdAtUtc: string;
  // Sync metadata
  syncedAtUtc: string | null;
  syncVersion: number;
}

// ===== Equipment Types =====

// Sync-compatible equipment status - matches Android sync types
export type EquipmentStatus = 'AVAILABLE' | 'CHECKED_OUT' | 'MAINTENANCE' | 'RETIRED';
// Sync-compatible equipment type - matches Android sync types
export type EquipmentType = 'TRAINING_MATERIAL';
export type ConflictStatus = 'PENDING' | 'RESOLVED' | 'CANCELLED';

export interface EquipmentItem {
  id: string;
  serialNumber: string;
  name: string;
  description: string | null;
  type: EquipmentType; // Sync field - maps to equipmentType in DB
  equipmentType?: EquipmentType; // DB field - for local queries
  status: EquipmentStatus;
  notes?: string | null;
  deviceId?: string; // Sync field
  createdByDeviceId?: string; // DB field
  syncVersion: number;
  createdAtUtc: string;
  modifiedAtUtc: string;
  syncedAtUtc: string | null;
}

export interface EquipmentCheckout {
  id: string;
  equipmentId: string;
  internalMemberId: string; // FK to Member.internalId
  /** @deprecated Use internalMemberId. Retained for backward compatibility. */
  membershipId: string | null;
  checkedOutAtUtc: string;
  checkedInAtUtc: string | null;
  checkedOutByDeviceId: string;
  checkedInByDeviceId: string | null;
  checkoutNotes: string | null;
  checkinNotes: string | null;
  conflictStatus: ConflictStatus | null;
  deviceId: string;
  syncVersion: number;
  createdAtUtc: string;
  modifiedAtUtc: string;
  syncedAtUtc: string | null;
}

// ===== Device Types =====

export type DeviceType = 'MEMBER_TABLET' | 'TRAINER_TABLET' | 'DISPLAY_EQUIPMENT' | 'DISPLAY_PRACTICE' | 'LAPTOP';

export interface DeviceInfo {
  id: string;
  name: string;
  type: DeviceType;
  lastSeenUtc: string | null;
  pairingDateUtc: string;
  ipAddress: string | null;
  port: number | null;
  isTrusted: boolean;
  isOnline: boolean;
}

// ===== Sync Types =====

export interface SyncConflict {
  id: string;
  conflictType: string;
  entityType: string;
  entityId: string;
  conflictingEntityId: string | null;
  localDeviceId: string;
  localDeviceName: string | null;
  localTimestamp: string;
  localSyncVersion: number;
  remoteDeviceId: string;
  remoteDeviceName: string | null;
  remoteTimestamp: string;
  remoteSyncVersion: number;
  status: 'PENDING' | 'RESOLVED' | 'SYNCED';
  resolution: string | null;
  resolvedByDeviceId: string | null;
  resolvedAtUtc: string | null;
  context: string | null;
  detectedAtUtc: string;
}

/**
 * Payload for incoming sync from tablets (full member data).
 */
export interface SyncPayloadIncoming {
  deviceId: string;
  deviceName: string;
  deviceType: DeviceType;
  schemaVersion: string;
  timestamp: string;
  members?: Member[];
  checkIns?: CheckIn[];
  practiceSessions?: PracticeSession[];
  scanEvents?: ScanEvent[];
  registrations?: NewMemberRegistration[];
  equipmentItems?: EquipmentItem[];
  equipmentCheckouts?: EquipmentCheckout[];
}

/**
 * Payload for outgoing sync to tablets (full member data per DD-9).
 * All member fields are now synced bidirectionally.
 */
export interface SyncPayloadOutgoing {
  deviceId: string;
  deviceName: string;
  deviceType: DeviceType;
  schemaVersion: string;
  timestamp: string;
  members?: Member[];
  checkIns?: CheckIn[];
  practiceSessions?: PracticeSession[];
  scanEvents?: ScanEvent[];
  registrations?: NewMemberRegistration[];
  equipmentItems?: EquipmentItem[];
  equipmentCheckouts?: EquipmentCheckout[];
}

/** @deprecated Use SyncPayloadIncoming or SyncPayloadOutgoing */
export interface SyncPayload {
  deviceId: string;
  deviceName: string;
  deviceType: DeviceType;
  schemaVersion: string;
  timestamp: string;
  members?: Member[];
  checkIns?: CheckIn[];
  practiceSessions?: PracticeSession[];
  scanEvents?: ScanEvent[];
  registrations?: NewMemberRegistration[];
  equipmentItems?: EquipmentItem[];
  equipmentCheckouts?: EquipmentCheckout[];
}

// ===== Member Utility Functions =====

/**
 * Calculate age from a birth date string.
 * @param birthDate ISO date string (YYYY-MM-DD) or null
 * @returns Age in years, or null if birthDate is invalid
 */
export function calculateAge(birthDate: string | null): number | null {
  if (!birthDate) return null;

  try {
    const birth = new Date(birthDate);
    if (isNaN(birth.getTime())) return null;

    const today = new Date();
    let age = today.getFullYear() - birth.getFullYear();
    const monthDiff = today.getMonth() - birth.getMonth();

    // Adjust age if birthday hasn't occurred yet this year
    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
      age--;
    }

    return age;
  } catch {
    return null;
  }
}

/**
 * Check if a member is an adult (age >= 18).
 * @param member The member to check
 * @returns true if adult, false if minor or age cannot be determined
 */
export function isAdult(member: Member): boolean {
  const age = calculateAge(member.birthDate);
  return age !== null && age >= 18;
}

/**
 * Check if a member needs an ID photo (adult TRIAL member without ID photo).
 * FULL members don't need ID verification - only TRIAL members do.
 * @param member The member to check
 * @returns true if adult TRIAL member and missing ID photo
 */
export function needsIdPhoto(member: Member): boolean {
  return isAdult(member) && member.memberLifecycleStage === 'TRIAL' && !member.idPhotoPath;
}

/** Status of ID photo requirement */
export type IdPhotoStatus = 'available' | 'pending' | 'not_required';

/**
 * Get the ID photo status for a member.
 * @param member The member to check
 * @returns 'available' if ID photo exists, 'pending' if adult TRIAL member needs ID, 'not_required' for minors/FULL members
 */
export function getIdPhotoStatus(member: Member): IdPhotoStatus {
  if (!isAdult(member)) return 'not_required';
  if (member.memberLifecycleStage === 'FULL') return 'not_required';
  if (member.idPhotoPath) return 'available';
  return 'pending';
}
